### ISSUE TITLE
Build error: ROCMHeaderWrapper not found after #1454

### ISSUE THREAD

**Issue by atamazov** [ROCmSoftwarePlatform/MIOpen#1473]
MIOpen build fails after #1454. Tried with vanilla ROCm 5.0 on two machines + rocm-cmake installed.

rocm-cmake installation/update:
```bash
$ cd ~/github
$ git clone https://github.com/RadeonOpenCompute/rocm-cmake
$ cd rocm-cmake
$ git checkout 4a1514850195360ce772182d34b6b5afd8f37253
$ mkdir build
$ cd build && cmake ..
$ sudo cmake --build . --target install
$ ls -l /opt/rocm/share/rocm/cmake
...
-rw-r--r-- 1 root root  5294 Mar 23 21:38 ROCMHeaderWrapper.cmake
...
```

MIOpen build:
```bash
atamazov@morpheus:~/github/MLOpen1/build/release.hip-clang$ CXX=/opt/rocm/llvm/bin/clang++ cmake -DBUILD_DEV=On -DCMAKE_BUILD_TYPE=release -DMIOPEN_USE_MLIR=Off ../..
...
CMake Error at CMakeLists.txt:61 (include):
  include could not find load file:

    ROCMHeaderWrapper
```

@junliume @pfultz2 @arvindcheru :red_circle: This seems to be a blocker issue for me.

**Comment by atamazov** [comment]
Works on third machine... Maybe this is my local configuration issue. Closing until issue is clarified.

**Comment by junliume** [comment]
> Works on third machine... Maybe this is my local configuration issue. Closing until issue is clarified.

@atamazov MIOpen dependency needs to be updated (rocm-cmake specifically). I've met this issue once and re-ran the dependency installation fixed it. Let me know if this issue persists on your end. :)

**Comment by atamazov** [comment]
The reason was another, outdated, instance of rocm-cmake installed on my machine due to some reason, at `/usr/local/share/rocm/cmake`. Removal of it resolves the problem.

**Comment by junliume** [comment]
Need to reopen this issue due to break of backward compatibility:

@pfultz2 @arvindcheru please advice.

[Steps to reproduce]: pull any 5.0 docker and try to build MIOpen within it.

**Comment by atamazov** [comment]
@junliume 
> pull any 5.0 docker and try to build MIOpen within it.

This is normal because 5.0 contains outdated dependencies (including rocm-cmake). Further releases will contain updated rocm-cmake because [requirements.txt is updated in #1454](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1454/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552).

> we should remove or replace /opt/rocm/share/rocm/cmake but it needs to be documented...

The problem only affects developers who want to build fresh MIOpen with older ROCm releases, _if they would like to skip_ the step described at https://github.com/ROCmSoftwarePlatform/MIOpen#installing-the-dependencies (**_as we usually do, BTW_**). They need to update rocm-cmake as shown in the [issue description](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1473#issue-1178535581). 

The users who follow the instructions won't encounter any problems. I think this ticket does not require any actions.

**Comment by arvindcheru** [comment]
Thanks for comments/inputs @junliume , @atamazov. 
As explained @atamazov in last comment, this dependency installation is done based on existing requirements.txt processing,  as per our last changes, we did addition of rocm-cmake also to it. 
This was added, since the Wrapperfunction is only available from latest (0.7.3) releases of rocm-cmake. 

Will try to understand whether we can handle existing rocm-cmake by any means while do this installation.

**Comment by junliume** [comment]
> @junliume
> 
> > pull any 5.0 docker and try to build MIOpen within it.
> 
> This is normal because 5.0 contains outdated dependencies (including rocm-cmake). Further releases will contain updated rocm-cmake because [requirements.txt is updated in #1454](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1454/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552).
> 
> > we should remove or replace /opt/rocm/share/rocm/cmake but it needs to be documented...
> 
> The problem only affects developers who want to build fresh MIOpen with older ROCm releases, _and only if they skip_ the step described at https://github.com/ROCmSoftwarePlatform/MIOpen#installing-the-dependencies (**_as we usually do, BTW_**). They need to update rocm-cmake as shown in the [issue description](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1473#issue-1178535581).
> 
> The users who follow the instructions won't encounter any problems. I think this ticket does not require any actions.

@atamazov I ran the following script to install MIOpen, with the dependency installation:
> cd /root && git clone -b develop https://github.com/ROCmSoftwarePlatform/MIOpen && cd MIOpen && \
mkdir -p install_dir && **cmake -P install_deps.cmake --minimum --prefix /root/MIOpen/install_dir** && \
mkdir -p build && cd build && export CXX=/opt/rocm/llvm/bin/clang++ && \
cmake -DMIOPEN_BACKEND=HIP **-DCMAKE_PREFIX_PATH="/root/MIOpen/install_dir"** .. 

and still meet the above error, until I remove the /opt/rocm/share/rocm/cmake folder. It seems that `-DCMAKE_PREFIX_PATH="/root/MIOpen/install_dir"` is not taking proper priority over other default paths. Another approach I have tried is `cmake -P install_deps.cmake --minimum` without `--prefix` so that it is installed at `/usr/local`, but still the one at `/opt/roc/` got picked up and it is the outdated one :(

**Comment by arvindcheru** [comment]
One observation, currently build is done with rbuild for me where this issue is not see, it will do the dependency installation to the deps_hip directory and in turn will proceed with success build. 

@pfultz2,
 for now shall we add download like our previous version just an exception (for now) for rocm-cmake like we did https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1454/files/ef9930472b813515d1fd39db9b7dc24a8dd07989 in CMakeLists.txt.

**Comment by pfultz2** [comment]
> I see. /opt/rocm/share/rocm/cmake takes precedence over CMAKE_PREFIX_PATH.

So `/opt/rocm` get preferred over the system path because cmake first searchs `CMAKE_INSTALL_PREFIX`, which is `/opt/rocm`.

> One observation, currently build is done with rbuild for me where this issue is not see, it will do the dependency installation to the deps_hip directory and in turn will proceed with success build.

So we could update install_deps.cmake to use rbuild, I think it would be better, but I am not sure why it works over what we do currently. The difference here is that rbuild will set `CMAKE_PREFIX_PATH`, whereas with install_deps.cmake, users are relying on the dependencies being found in `/usr/local`. Either way, @atamazov seem to imply that `CMAKE_PREFIX_PATH` is not enough. 

Also, it seems we list rocm-cmake in two different files(dev-requirements.txt and requirements.txt), which could be a problem in the future(right now we are lucky they are using the same version).

> or now shall we add download like our previous version just an exception (for now) for rocm-cmake like we did 

No, we should not download dependencies at cmake time. However, we could set the minimum rocm-cmake version, this can find a correct rocm-cmake when there are multiple versions to be found: `find_package(ROCM 0.7.3 REQUIRED PATHS /opt/rocm)`. This might fix these issues.

**Comment by pfultz2** [comment]
So #1482 is a proposed fix.

**Comment by jinz2014** [comment]
Although rocm-cmake is installed, installing ROCm components still reports cmake errors (e.g. Unknown CMake command "rocm_package_setup_component).  Could you please comment how to troubleshoot the issue ? 


rocm-cmake/jammy,now 0.8.0.50400-72~22.04 amd64 [installed,automatic]
  rocm-cmake built using CMake

**Comment by arvindcheru** [comment]
> Although rocm-cmake is installed, installing ROCm components still reports cmake errors (e.g. Unknown CMake command "rocm_package_setup_component). Could you please comment how to troubleshoot the issue ?
> 
> rocm-cmake/jammy,now 0.8.0.50400-72~22.04 amd64 [installed,automatic] rocm-cmake built using CMake

Sorry for the delay in responding, 
I was unable to get same observation when I tried on recent build. Would like to know more details on environments/steps followed to get the observation shared. would appreciate opening another ticket if it is not related this issue with more details on the environment/source used like
Which version (is it a particular version onwards) of MIOpen this observation is found.
Would you mind sharing the build steps used, is it provided with provided Prefix path used

**Comment by atamazov** [comment]
@jinz2014 
- what is your rocm version? 
- you have to use rocm-cmake from rocm, see https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1473#issuecomment-1076741961

**Comment by jinz2014** [comment]
Thank you for your replies. Please ignore my issue.
